SimpleTraversability added.

JanWehrmann opened the pull request:
In addition to the main class the helper class
SimpleTraversabilityRadialLUT was added, which
is used in the postprocessing methods of
SimpleTraversability:

In addition Unit Tests were added for
SimpleTraversability and the neccessary changes
were made to the CMakeLists.

Changes in comparison to the envire version:
Constructors and Configs
- Setters and Getters for the SimpleTraversabilityConfig were added
  as well as a new constructor for the Config making the use of
  Configs more convenient.
- Postprocessing parameters now have a default value of 0 in all
  constructors.
- ObstacleClearance was addded to the config and the constructors
  enabling the use of growObstacles.

calculateTraversability
- Will now check if the value of an input differs from the
  input maps default value to determine UNKNOWN areas.

RadialLUT and closeNarrowPassages
- RadialLUT was moved to its own file and renamed
  SimpleTraversabilityRadialLUT.
- closeNarrowPassages will now use a copy of the map instead of
  using 255 as a marking value. These changes are also apparent
  in RadialLUT which does the marking.
- closeNarrowPassages will now use the whole RadialLUT instead
  of only one quadrant. A check was added to prevent access to
  cells outside of the grid.

saarnold commented on src/tools/SimpleTraversability.cpp:
```diff
@@ -0,0 +1,253 @@
+        }
+    }
+
+    std::swap(traversabilityGridOut, traversabilityGridIn);
```
The swap seems to be unnecessary. What about copying the grid temporary to look up the traversability class and directly modifying the input grid?
The same can be done in the growObstacles method.

JanWehrmann replied:
I agree, that would be better. I will make the change.

saarnold commented on src/tools/SimpleTraversability.hpp:
```diff
@@ -0,0 +1,192 @@
+     */
+    class SimpleTraversability {
+
+        enum INPUT_DATA {
```
Is this type used anywhere?

JanWehrmann replied:
No this is a remnant of how the inputs were handled before. I will remove it.

saarnold commented on src/tools/SimpleTraversability.hpp:
```diff
@@ -0,0 +1,192 @@
+            INPUT_COUNT
+        };
+
+    public:
```
Try to have only one private, protected and public scope per class definition.

JanWehrmann replied:
Ok. I will move the CLASSES enum into the second public scope then.

saarnold commented on src/tools/SimpleTraversability.hpp:
```diff
@@ -0,0 +1,192 @@
+                double minPassageWidth = 0,
+                double obstacleClearance = 0);
+
+        void setConfig(const SimpleTraversabilityConfig& config);
```
Not mandatory to change it now, but if there already is a config struct (like SimpleTraversabilityConfig) that should be sufficient to initialize or change the config.

JanWehrmann replied:
I added the function since the config is private (not sure if that is necessary, I left it like it was in the original) one was unable to change the config if for some reason the default constructor was used.

saarnold commented on src/tools/SimpleTraversability.hpp:
```diff
@@ -0,0 +1,192 @@
+         * Note that the default value of the input maps will be
+         * used to determine UNKNOWN cells.
+         */
+        bool calculateTraversability(
```
calculateTraversability, closeNarrowPassages and growObstacles could be const right?

JanWehrmann replied:
Yes.

saarnold commented on src/tools/SimpleTraversabilityRadialLUT.hpp:
```diff
@@ -0,0 +1,68 @@
+        * Checks for the targetedValue within the in distance around the given center and calls markSingleRadius on them.
+        */
+        void markAllRadius(grid::TraversabilityGrid& traversabilityGrid, int centerX, int centerY, int targetedValue);
+
```
markAllRadius and markSingleRadius are const as well right?

JanWehrmann replied:
Yes.

saarnold commented on src/tools/SimpleTraversabilityRadialLUT.hpp:
```diff
@@ -0,0 +1,68 @@
+    /**
+     * Radial lookup table used in SimpleTraversability::closeNarrowPassages.
+    **/
+    struct SimpleTraversabilityRadialLUT
```
This should be a class and it's members protected or private

JanWehrmann replied:
I will make the changes.

Commit:
SimpleTraversability added.

In addition to the main class the helper class
SimpleTraversabilityRadialLUT was added, which
is used in the postprocessing methods of
SimpleTraversability:

In addition Unit Tests were added for
SimpleTraversability and the neccessary changes
were made to the CMakeLists.

Changes in comparison to the envire version:
Constructors and Configs
- Setters and Getters for the SimpleTraversabilityConfig were added
  as well as a new constructor for the Config making the use
  of Configs more convenient.
- Postprocessing parameters now have a default value of 0 in all
  constructors.
- ObstacleClearance was addded to the config and the constructors
  enabling the use of growObstacles.

calculateTraversability
- Will now check if the value of an input differs from the
  input maps default value to determine UNKNOWN areas.

RadialLUT and closeNarrowPassages
- RadialLUT was moved to its own file and renamed
  SimpleTraversabilityRadialLUT.
- closeNarrowPassages will now use a copy of the map instead of
  using 255 as a marking value. These changes are also apparent
  in RadialLUT which does the marking.
- closeNarrowPassages will now use the whole RadialLUT instead
  of only one quadrant. A check was added to prevent access to
  cells outside of the grid.
